### PR TITLE
Prevent strong chance of failed physical point inversion

### PR DESCRIPTION
**Pull request by lindsayad**
Right now when someone calls `FEProblemBase::reinitElemPhys` with a
vector of physical points, we will attempt to invert those physical
points into the reference element space for both undisplaced and
displaced meshes if the latter exists. It is not difficult to imagine
that for a given set of physical points you may easily fall off the
displaced mesh for large enough displacements/deformations. Let's see if
this makes anything fail

Refs #10823

### CODE REVIEW

**Comment by tophmatthews** [comment]
Need to comment out `suppress_displaced_init` in the call

**Comment by lindsayad** [comment]
If this `reinitElemPhys` patch passes all tests maybe we can wait to do `reinitNeighborPhys` in a later PR. Does that sound good @tophmatthews ?

File: framework/src/problems/FEProblemBase.C [hunk 1]
```diff
@@ -1735,7 +1735,7 @@ void
 FEProblemBase::reinitElemPhys(const Elem * elem,
                               const std::vector<Point> & phys_points_in_elem,
                               THREAD_ID tid,
-                              bool suppress_displaced_init)
```
Comment (loganharbour) [inline framework/src/problems/FEProblemBase.C:1738]: This isn't used in the DisplacedProblem method either - can we remove it? Or are there other applications that still pass something in here?

**Comment by tophmatthews** [comment]
Whats up with this?

**Comment by lindsayad** [comment]
Divided human resource :smile:

**Comment by lindsayad** [comment]
Trying to get it through though! Made some progress...

**Comment by lindsayad** [comment]
This is a nightmare. A total nightmare. The `Constraint` system and its associated logic in `NonlinearSystemBase` never ceases to horrify me. It's the single biggest time-suck trying to sort through these problems. Never-ending streams of coding without encapsulating. Always going through `_fe_problem`.... Aaaaaaaaaaaaaaaaaaaarrrrrrrrrrrrrrrrrrrgggggggggggggggghhhhhhhhhhhhhhhhh

**Comment by lindsayad** [comment]
Oh it's looking like a great success!

**Comment by lindsayad** [comment]
@tophmatthews have you tried this on your problematic case?

**Comment by lindsayad** [comment]
If it works, can you review/approve this?

**Comment by tophmatthews** [comment]
> If it works, can you review/approve this?

I've been staying away due to all the red in civet, but it looks good now, let me test it.

**Comment by lindsayad** [comment]
@loganharbour unless @tophmatthews comes back with bad news, I think this is ready to review/go

**Comment by tophmatthews** [comment]
Ugg...I'm trying to reproduce...taking longer than I thought...

File: framework/src/problems/DisplacedProblem.C [hunk 2]
```diff
@@ -617,9 +617,11 @@ DisplacedProblem::reinitElem(const Elem * elem, THREAD_ID tid)
 void
 DisplacedProblem::reinitElemPhys(const Elem * elem,
                                  const std::vector<Point> & phys_points_in_elem,
-                                 THREAD_ID tid,
-                                 bool)
+                                 THREAD_ID tid)
 {
+  mooseAssert(_mesh.getMesh().query_elem_ptr(elem->id()) == elem,
+              "Are you calling this method with a undisplaced mesh element?");
```
Comment (loganharbour) [inline framework/src/problems/DisplacedProblem.C:623]: Beautiful assertions 🙌🏼 

FYI in the future for even less code, we have the `MooseMesh::queryElemPtr()` API

File: framework/src/problems/FEProblemBase.C [hunk 3]
```diff
@@ -5002,6 +4992,10 @@ FEProblemBase::solve()
 {
   TIME_SECTION(_solve_timer);
 
+  clearAllDofIndices();
```
Comment (loganharbour) [inline framework/src/problems/FEProblemBase.C:4995]: I think the reasoning is clear, but I think it's still a good idea to put a comment here as to why you ended up needing to do this

File: framework/src/systems/NonlinearSystemBase.C [hunk 4]
```diff
@@ -981,6 +981,12 @@ NonlinearSystemBase::setConstraintSecondaryValues(NumericVector<Number> & soluti
   std::map<std::pair<unsigned int, unsigned int>, PenetrationLocator *> * penetration_locators =
       NULL;
 
+  mooseAssert(displaced ? bool(_fe_problem.getDisplacedProblem().get()) : true,
+              "If we're calling this method with displaced = true, then we better well have a "
+              "displaced problem");
```
Comment (loganharbour) [inline framework/src/systems/NonlinearSystemBase.C:986]:
```suggestion
  if (displaced)
    mooseAssert(_fe_problem.getDisplacedProblem().get(),
               "If we're calling this method with displaced = true, then we better well have a "
               "displaced problem");
```

With the assertion opt being:

```
if (displaced)
  (void)0
```

this is fine and will be removed by the compiler. I used to do this, but the ternary assertion with a true just looks like hot trash IMO

File: framework/src/systems/NonlinearSystemBase.C [hunk 5]
```diff
@@ -1115,6 +1121,12 @@ NonlinearSystemBase::constraintResiduals(NumericVector<Number> & residual, bool
...
```
Comment (loganharbour) [inline framework/src/systems/NonlinearSystemBase.C:1126]: Same here

File: framework/src/systems/NonlinearSystemBase.C [hunk 6]
```diff
@@ -1852,6 +1864,12 @@ NonlinearSystemBase::constraintJacobians(bool displaced)
...
```
Comment (loganharbour) [inline framework/src/systems/NonlinearSystemBase.C:1869]: And here

File: framework/src/systems/SystemBase.C [hunk 7]
```diff
@@ -1549,6 +1549,13 @@ SystemBase::jacobianSetup()
     _vars[tid].jacobianSetup();
 }
 
+void
+SystemBase::clearAllDofIndices()
+{
+  for (const auto tid : make_range(libMesh::n_threads()))
+    _vars[tid].clearAllDofIndices();
```
Comment (loganharbour) [inline framework/src/systems/SystemBase.C:1556]:
```suggestion
  for (auto & var_warehouse : _vars)
    var_warehouse.clearAllDofIndices();
```

Why not?

**Review by loganharbour — CHANGES_REQUESTED** [review]
All code style requests, I think.

Removing the extra calls from the reinit methods in FEProblemBase makes this read SO much better

**Review by loganharbour — APPROVED** [review]

**Comment by lindsayad** [comment]
@tophmatthews how you doing over there?

**Comment by tophmatthews** [comment]
I'm trying...I have so many balls in the air, I can't find a reproducible case that fails obviously to really test this. I'll keep looking, but if tests pass then, don't wait for me...

**Comment by tophmatthews** [comment]
Wow, it took forever, but yes, can confirm 👍 

You da 💣  @lindsayad

**Pull request merged**